### ISSUE TITLE
Division is definable and useful already in Semirings

### ISSUE THREAD

**Issue by hseg** [chessai/semirings#62]
```divide``` should be a method of ```Semiring```, as it makes sense from the moment multiplication forms a monoid.
In particular, am currently working on the localization of an integral domain away from a prime. To normalize fractions, I need to be able to compute decompositions ```x=p^k*m```. This doesn't need gcds to be uniquely definable in general, just at powers of a fixed prime, which they always are.
Incidentally, this decomposition is frequently useful, and can probably be given instance-specific speedups, so I recommend it also be added to the class, with default definition
```
decomp :: Semiring a => a -> a -> (Natural,a)
decomp p x = iterateM step (0, x)
 where
  step (k, m) = (k+1 ,) `fmap` (m `divide` p)
  iterateM f !acc = maybe !acc (iterateM f) (f acc)
```

**Comment by Bodigrim** [comment]
What is the semantics of `divide` in a `Semiring`, which is not an integral domain? 

I thought about introducing an intermediate typeclass `IntegralDomain` between `Semiring` and `GcdDomain`, but decided against it because of lack of use cases of data, which is `IntegralDomain`, but not `GcdDomain`. I'm open to rethink this decision. Could you possibly share a data type you are using to model "the localization of an integral domain away from a prime"?

**Comment by hseg** [comment]
Fair. The laws as written:
```
\x y -> (x * y) `divide` y == Just x
\x y -> maybe True (\z -> x == z * y) (x `divide` y)
```
fail for `y` a divisor of `0`.

To account for such `y`, we should instead have that
```
x*y == z*y => (x*y) `divide` y == (z*y) `divide` y
maybe x (*y) (x `divide` y) == x
```
in particular, ``fmap ((*y) . (minus x)) ((x*y) `divide` y) == Just 0``, but you
have no guarantee that ``(`divide` y) . (*y) == Just``.

As for my localization example, here is a snippet:
```haskell
newtype m `Loc` r = Loc (r, m)

instance (Action m r, Eq r) => Eq (m `Loc` r) where
  (Loc (x, m)) == (Loc (y, n)) = n `act` x == m `act` y

instance (Action m r, Ord r) => Ord (m `Loc` r) where
  (Loc (x, m)) `compare` (Loc (y, n)) = (n `act` x) `compare` (m `act` y)

instance (Show m, Show r) => Show (m `Loc` r) where
  show (Loc (x, m)) = show x ++ "/" ++ show m

instance Functor (Loc m) where
  fmap f (Loc (x, m)) = Loc (f x, m)

instance (Action m r, Monoid m, Semiring r) => Semiring (m `Loc` r) where
  (Loc (x, m)) `plus` (Loc (y, n)) = Loc (n `act` x + m `act` y, m <> n)
  (Loc (x, m)) `times` (Loc (y, n)) = Loc (x * y, m <> n)
  fromNatural = inj . fromNatural @r

instance (Action m r, Monoid m, Ring r) => Ring (m `Loc` r) where
  negate (Loc (a, m)) = Loc (negate a, m)

type R₂ a = AwayFrom (FromNatural 2 a) `Loc` a

newtype AwayFrom p = Exp N
  deriving stock (Eq, Ord)
  deriving (Semigroup, Monoid) via (Add N)

data FromNatural (n :: Nat) r

instance (KnownNat n, Semiring r) => Reifies (FromNatural n r) r where
  reflect = fromNatural $ fromIntegral (reflect (Proxy @n))

instance (Reifies p s, Show s) => Show (AwayFrom p) where
  show (Exp i) = show (reflect (Proxy @p)) ++ "^" ++ show i

instance Semiring r => (AwayFrom (FromNatural n r) `Loc` r) `Inverts` n where
  inv' _ = Loc (one, Exp one)

instance (Semiring r, Reifies t r) => Action (AwayFrom t) r where
  act (Exp e) = ((reflect (Proxy @t) ^ e) *)
```

Now consider trying to express the fact that in `R₂ Int`, all numbers of form
`±2ᵏ` are invertible. You'd want to decompose a potential unit as `2ᵏm` and
check `m == 1`. But that's currently impossible with the `Semiring` API.

Of course, in this case `Int` is a `GcdDomain`, but I don't like
overconstraining my code.

Hope this clarifies things.

**Comment by hseg** [comment]
OK, a naive attempt shows the following problem points:
easy instances
* Mod2 -- is using the default Euclidean instance, but could be given explicitly
* Proxy
* Bool
* (Maybe a)
* (IO a)
* (Dual a)
* (Const a b)
* (Ap f a)
* (Fixed a) -- seems we can just use the Fractional instance

can make an instance up iso to Int, which seems like the right semantics
* CSpeed
* CCc
* CClock
* CTime
* CUSeconds
* CSUSeconds

probably can be made to work, needs research
* (Set a)
* (IntSetOf a)
* (Map k v)
* (IntMapOf k v)
* (HashSet a)
* (HashMap k v)

problems
* (WrappedNum a) -- would need integral/fieldlike split
* (a -> b) -- problem: need to catch indivisibility at any point

suggestion: instead of `divide`, expose `decomp :: a -> a -> (Natural, a)` and
let `divide x y = fst (decomp x y) /= 0`

**Comment by hseg** [comment]
This alternative also is no good. Fundamentally, the problem is that for the `a -> b` instance, we need to be able to evaluate ``p `divides` x=forall $ \i -> p i `divides` x i``, for some `forall :: (a -> Bool) -> Bool`. But afaik, this is impossible in general, even with laziness tricks. So we're left with either killing the `a -> b` instance, restricting it to finite `a`, or splitting the Semiring class into decidable/undecidable semirings, with unclear semantics. Or we could specialise `divide` to integral domains, since product rings are never integral.

**Comment by chessai** [comment]
I am in the middle of moving across the United States, so I will not be
very available for the next week or so.

Thanks

**Comment by hseg** [comment]
Checked what `algebra` does here - it restricts `divides` to integral domains, which rules out the problematic instances we have here (`Num` makes no promises of being integral). Can live with the reduced generality, or we can recover it by naming the integral class `DecidableDivision` or the like.
Writing a quick patch for integral domains, if generality is desired it will be a small edit on top of that.

**Comment by hseg** [comment]
BTW, `algebra` also distinguishes between `Euclidean.divide :: a -> a -> (a,a)`, `Integral.divides :: a -> a -> Bool` and `Integral.MaybeQuot :: a -> a -> Maybe a`.
Given that we diverge on the name for `Euclidean`, we might as well steal `divide` for `:: a -> a -> Maybe a`, possibly adding some combinators on top of it (eg `divides`, `dividesExact`, `divDecomp`).

**Comment by Bodigrim** [comment]
> As for my localization example, here is a snippet:

If I'm not mistaken, basically we are talking about 

```haskell
import GHC.TypeNats 

data Loc a (p :: Nat) = Loc { numer :: a, denomExp :: Int }

instance (Real a, KnownNat p) => Real (Loc a p) where 
  toRational x@(Loc a k) = toRational a * (1 % toInteger (natVal x) ^ k)
```

and you'd like to define

```haskell
isUnit :: (KnownNat p, Semiring a) => Loc a p -> Bool
```

Is my understanding correct?

**Comment by hseg** [comment]
Better:
```
invert :: (KnownNat p, Semiring a) => Loc a p -> Maybe (Loc a p)
invert (Loc a k) = let (k', m) = decomp (natVal p) a in fmap (`Loc` (k-k')) (one `divide` m) 
```

**Comment by Bodigrim** [comment]
One can define
```haskell
instance (GcdDomain a, KnownNat p) => GcdDomain (Loc a p) where 
  gcd (Loc a _) (Loc b _) = Loc (gcd a b) 0 
  lcm (Loc a _) (Loc b _) = Loc (lcm a b) 0 
  divide (Loc a k) (Loc b l) = (`Loc` (k - l)) <$> divide a b
```
but you do not want to impose `GcdDomain a`, right? Why? Do you have a concrete example of `a`, which is `IntegralDomain`, but not `GcdDomain`?

I very much understand the desire to constrain types as little as possible, just for the sake of mathematical purity, and I can certainly relate. However, this is an avenue to 100+ type classes as in [`algebra`](http://hackage.haskell.org/package/algebra). This is why I insist on having concrete examples of constructive structures, which require introducing additional type classes.

For instance, `Euclidean` was brought to life to cover Gaussian/Eisenstein integers, which cannot be `Integral` because of no `toInteger` function. And further `GcdDomain` was separated from `Euclidean` because [polynomials](https://hackage.haskell.org/package/poly-0.4.0.0/docs/Data-Poly.html#t:GcdDomain) over `Euclidean` coefficients are not `Euclidean`.

**Comment by hseg** [comment]
Not in my current code, no. But e.g. quadratic extensions aren't all gcd domains, yet are still integral. Will probably bump into wanting those soon, and in particular wanting to take their localizations at particular elements.
Am afk, will edit with code sample later.

On August 19, 2020 3:21:51 AM GMT+03:00, Bodigrim <notifications@github.com> wrote:
>One can define
>```haskell
>instance (GcdDomain a, KnownNat p) => GcdDomain (Loc a p) where 
>  gcd (Loc a _) (Loc b _) = Loc (gcd a b) 0 
>  lcm (Loc a _) (Loc b _) = Loc (lcm a b) 0 
>  divide (Loc a k) (Loc b l) = (`Loc` (k - l)) <$> divide a b
>```
>but you do not want to impose `GcdDomain a`, right? Why? Do you have a
>concrete example of `a`, which is `IntegralDomain`, but not
>`GcdDomain`?
>
>I very much understand the desire to constrain types as little as
>possible, just for the sake of mathematical purity, and I can certainly
>relate. However, this is an avenue to 100+ type classes as in
>[`algebra`](http://hackage.haskell.org/package/algebra). This is why I
>insist on having concrete examples of constructive structures, which
>require introducing additional type classes.
>
>For instance, `Euclidean` was created to cover Gaussian/Eisenstein
>integers, which cannot be `Integral` because of no `toInteger`
>function. And further `GcdDomain` was separated from `Euclidean`
>because
>[polynomials](https://hackage.haskell.org/package/poly-0.4.0.0/docs/Data-Poly.html#t:GcdDomain)
>over `Euclidean` coefficients are not `Euclidean`. 
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/chessai/semirings/issues/62#issuecomment-675782054

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by hseg** [comment]
OK, thinking on it a bit more has clarified what I need. In my code, I basically have:
```haskell
data Loc r e = Loc { num :: r, log_denom :: Natural }
data family Unit r
newtype instance Unit (Loc r e) = ULoc (Unit r, Integer)

checkUnit :: (Semiring r, Unital r, Reifies e r)
          => Loc r e -> Maybe (Unit (Loc r e))
checkUnit (Loc n ld) =
    let (k, m) = n `divDecomp` reflect (Proxy @e)
    in (,fromIntegral ld - k) <$> checkUnit m
```
Which unless we have `divDecomp :: Semiring r => r -> r -> (Natural, r)` won't work.

My earlier assertions that this patch is unnecessary only holds when `GcdDomain r`, which is the case in my code right now, but won't hold if I decide to use it e.g. in code needing `ℤ[√5]₂`.

**Comment by Bodigrim** [comment]
It happened to me to implement 

```haskell
isUnit :: GcdDomain a => a -> Bool
isUnit d = isJust (one `divide` d)
```

I've been often thinking to make `isUnit` a member of `Semiring` type class with a default implementation in terms of `GcdDomain`. @hseg would it be enough for your applications?

**Comment by hseg** [comment]
No -- in my code I need type-level knowledge of `isUnit`:
```haskell
data family Unit r
class Group (Unit r) => Unital r where
  injU :: Unit r -> r
  assertUnit :: HasCallStack => r -> Unit r
  assertUnit = fromJust . checkUnit
  checkUnit :: r -> Maybe (Unit r)
  checkUnit = Just . assertUnit
```
so that I can write `sub :: Unital r => Monomial n -> Vector n (Unit r) -> Unit r`
(`Monomial` here is a Laurent monomial, ie `~ Vector n Int`). This gives me `eval :: (Unital r, Ring r) => Laurent n r -> Vector n (Unit r) -> r`.

As noted, for the exact problem I'm working on right now I have:
```haskell
data Loc r e = Loc { num :: r, log_denom :: Natural }
newtype instance Unit (Loc r e) = ULoc (Unit r, Sum Integer)
    deriving newtype (Semigroup, Monoid, Group)

-- | Note: Only legal for @e@ prime. Otherwise, iterate @Loc@ at prime divisors
-- of @e@
instance (Semiring r, Unital r, Reifies e r) => Unital (Loc r e) where
    injU (ULoc (ur, Sum ld))
      | ld >= 0 = (injU ur * reflect (Proxy @e) ^ ld) `Loc` 0
      | ld < 0 = injU ur `Loc` fromInteger (negate ld)
    checkUnit (Loc n ld) =
        let (k, m) = n `divDecomp` reflect (Proxy @e)
        in (,fromIntegral ld - k) <$> checkUnit m
```

Now, it turns out that in my current project I only ever need `Loc r e` for `Euclidean r`, so I *could* specialize this instance to `Euclidean r`.
However, if I end up needing algebraic integers, I will need this generality.
Of course, I could refactor this API by insisting that instances provide their own smart constructors for `Unit`. This is a somewhat reasonable position for `assertUnit`, since clients presumably have informal proofs there that the element in question is a unit, but uses of `checkUnit` won't be able to adapt.
Again, though, this isn't an issue in my current project, since I only ever use `assertUnit`, and at statically-known values at that.

TL;DR: Proposed compromise wouldn't help, can avoid needing this issue in my current project at the cost of some inelegance, either in instance or in API, but a future project might need this. Am OK archiving this issue until an actual use case comes up.

PS: Note for whoever ends up implementing this patch: the testsuite assumes <https://github.com/andrewthad/quickcheck-classes> is in sync with our API, so needs to be updated as well. Have done some work in this direction, but some laws need to be hashed out for that.

**Comment by chessai** [comment]
Closing for now, feel free to re-open if there's any further work
